<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2

In [2]:
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [3]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [4]:
# текст запроса
query_3_1 = f'''SELECT COUNT(id) 
                FROM public.vacancies'''

In [5]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df

C:\Users\Maria\AppData\Local\Temp\ipykernel_4020\1114379196.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [6]:
# текст запроса
query_3_2 = f'''SELECT COUNT(id) 
                FROM public.employers'''

In [7]:
# результат запроса
df = pd.read_sql_query(query_3_2, connection)
df

C:\Users\Maria\AppData\Local\Temp\ipykernel_4020\3062646138.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [8]:
# текст запроса
query_3_3 = f'''SELECT COUNT(id) 
                FROM public.areas'''

In [9]:
# результат запроса
df = pd.read_sql_query(query_3_3, connection)
df

C:\Users\Maria\AppData\Local\Temp\ipykernel_4020\928149064.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [10]:
# текст запроса
query_3_4 = f'''SELECT COUNT(id)
                FROM public.industries'''

In [11]:
# результат запроса
df = pd.read_sql_query(query_3_4, connection)
df

C:\Users\Maria\AppData\Local\Temp\ipykernel_4020\2101691108.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


count
0    294

23501 работодатель предлагает 49197 вакансий в 1362 регионах. По количеству регионов можно сделать вывод, что в качетве регионов могут быть города, деревни как РФ, так и других стран (позже узнаем, что за регион страна тоже считается). Количество сфер деятельности 294. Среди работодателей есть как государственные учреждения, так и частные компании и ИП.

***

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [12]:
# текст запроса
query_4_1 = f'''SELECT a.name cities, 
                    COUNT(v.id) cnt_vacancies
                FROM public.vacancies v 
                    LEFT JOIN public.areas a on a.id = v.area_id
                GROUP BY cities
                ORDER BY cnt_vacancies DESC
                LIMIT 5
                '''

In [13]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\Maria\AppData\Local\Temp\ipykernel_4020\2457173557.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


cities  cnt_vacancies
0           Москва           5333
1  Санкт-Петербург           2851
2            Минск           2112
3      Новосибирск           2006
4           Алматы           1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [14]:
# текст запроса
query_4_2 = f'''SELECT COUNT(v.id) cnt_vacancies
                FROM public.vacancies v
                WHERE salary_from IS NOT NULL 
                    OR salary_to IS NOT NULL
                '''

In [15]:
# результат запроса
df = pd.read_sql_query(query_4_2, connection)
df

C:\Users\Maria\AppData\Local\Temp\ipykernel_4020\970655580.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


cnt_vacancies
0          24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [16]:
# текст запроса
query_4_3_1 = f'''SELECT ROUND(AVG(salary_from)) avg_salary_from, 
                    ROUND(AVG(salary_to)) avg_salary_to
                FROM public.vacancies v
                '''

In [17]:
# результат запроса
df = pd.read_sql_query(query_4_3_1, connection)
df

C:\Users\Maria\AppData\Local\Temp\ipykernel_4020\248808154.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3_1, connection)


avg_salary_from  avg_salary_to
0          71065.0       110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [18]:
# текст запроса
query_4_4 = f'''SELECT v.schedule, 
                    v.employment,
                    COUNT(name) count
                FROM public.vacancies v
                GROUP BY v.schedule, v.employment
                ORDER BY count DESC
                '''

In [19]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\Maria\AppData\Local\Temp\ipykernel_4020\2022935265.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


schedule           employment  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [20]:
# текст запроса
query_4_5 = f'''SELECT v.experience, 
                    COUNT(name) count
                FROM public.vacancies v
                GROUP BY v.experience
                ORDER BY count
                '''

In [21]:
# результат запроса
df = pd.read_sql_query(query_4_5, connection)
df

C:\Users\Maria\AppData\Local\Temp\ipykernel_4020\2316260449.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

Результаты запросов:
1. Больше всего свободных вакансий в Москве, на втором месте СПб.
2. Из первого пункта мы узнали, что всего представлено 49197 вакансий, из них всего в 24073 заполнено хотя бы одно поле с зарплатой, можно предположить, что большинство работодателей не хотят указывать этот пункт.
3. Средняя вилка заработной платы начинается от 71.065 руб. и заканчивается 110.537 руб.
4. Сочетание типа "полный день - полная занятость" имеет наибольшее число вакансий, следующим идет сочетание типа "удаленная работа - полная занятость", на последнем месте - сочетание типа "сменный график - проектная работа", сочетание такого типа имеет всего одна вакансия.
5. Больше всего вакансий с требуемым стажем от 1 года до 3 лет, меньше всего со стажем более 6 лет. Предположительно, такой стаж работы в основном требуется на руководящих должностях, поэтому их сравнительно немного.

Понятно, что большинство свободных вакансий будет размещаться больших городах (особеннов в Москве и СПб), так как там располагаются многие офисы крупных компаний. Из-за этого и такая высокая средняя вилка заработной платы. По результатам запроса было выяснено, что работодатель не стремится указывать предположительную заработную плату, ведь часто на собеседованиях про желаемую заработную плату спрашивают, от чего в последствие и могут отталкиваться при устройстве на работу, ведь кандидат может занизить свои требования, что выгодно работодателю. Работодатели, в основном, ищут соискателя на полный рабочий день, сочетание типа "удаленная работа - полная занятость" на втором месте говоит о том, что работодатель стремится оптимизировать расходы путем отказа от лишних расходов на аренду рабочих площадей. Большой плюс удаленной работы для работодателя еще состоит в том, что соискатель может находиться в любой точке планеты (с доступом к интернету), поэтому можно привлекать отличных специалистов из других стран и регионов. Запрос, касающийся количества вакансий в зависимости от опыта работы, показал, что 53 % вакансий в графе "требуемый опыт" имеют значение "от 1 года до 3 лет", предполагаю, что работодатели просто ставят нижнюю планку с опытом, так как не хотят брать рекрутера без опыта вообще (вакансии с графой "нет опыта" составляют всего 14,6 % от общего количества вакансий).

***

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [22]:
# текст запроса
query_5_1 = f'''SELECT e.name, 
                    COUNT(v.employer_id) count
                FROM public.vacancies v
                    LEFT JOIN public.employers e ON v.employer_id = e.id
                GROUP BY e.name
                ORDER BY count DESC
                LIMIT 5
                '''

In [23]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\Maria\AppData\Local\Temp\ipykernel_4020\3834399874.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


name  count
0         Яндекс   1933
1     Ростелеком    491
2       Тинькофф    444
3           СБЕР    428
4  Газпром нефть    331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [24]:
# текст запроса
query_5_2 = f'''SELECT a.name area_name,
                    COUNT(e.id) cnt_employers,
                    COUNT(v.id) cnt_vacancies
                FROM  public.areas a LEFT JOIN public.vacancies v ON a.id = v.area_id 
                    LEFT JOIN public.employers e ON a.id = e.area
                WHERE v.id IS NULL
                GROUP BY a.id
                ORDER BY cnt_employers DESC
                LIMIT 10
            '''

In [25]:
# результат запроса
df = pd.read_sql_query(query_5_2, connection)
df

C:\Users\Maria\AppData\Local\Temp\ipykernel_4020\2250345976.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


area_name  cnt_employers  cnt_vacancies
0                 Россия            410              0
1              Казахстан            207              0
2     Московская область             75              0
3     Краснодарский край             19              0
4               Беларусь             18              0
5     Ростовская область             18              0
6            Азербайджан             17              0
7  Нижегородская область             16              0
8   Республика Татарстан             16              0
9             Узбекистан             15              0

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [26]:
# текст запроса
query_5_3 = f'''SELECT e.name, 
                    COUNT(distinct v.area_id)
                FROM public.vacancies v
                    LEFT JOIN public.employers e ON v.employer_id = e.id
                GROUP BY e.name
                ORDER BY 2 DESC
                LIMIT 10
                '''

In [27]:
# результат запроса
df = pd.read_sql_query(query_5_3, connection)
df

C:\Users\Maria\AppData\Local\Temp\ipykernel_4020\201667002.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


name  count
0                      Яндекс    181
1                  Ростелеком    152
2                  Спецремонт    116
3      Поляков Денис Иванович     88
4                   ООО ЕФИН      71
5                  Совкомбанк     63
6                         МТС     55
7  ЭФКО, Управляющая компания     49
8                Почта России     48
9                        КРОН     48

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [28]:
# текст запроса
query_5_4 = f'''SELECT COUNT(e.id) cnt_employers
            FROM  public.employers e LEFT JOIN public.employers_industries ei ON e.id = ei.employer_id 
                LEFT JOIN public.industries i ON ei.industry_id = i.id
            WHERE i.id IS NULL
                '''

In [29]:
# результат запроса
df = pd.read_sql_query(query_5_4, connection)
df

C:\Users\Maria\AppData\Local\Temp\ipykernel_4020\3493105328.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


cnt_employers
0           8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [30]:
# текст запроса
query_5_5 = f'''SELECT e.name emp_name, 
                    COUNT(i.name) count_industry
                FROM public.employers e
                    LEFT JOIN public.employers_industries ei ON e.id = ei.employer_id
                    LEFT JOIN public.industries i ON ei.industry_id = i.id
                GROUP BY e.name
                HAVING COUNT(i.name)=4
                ORDER BY 1
                '''

In [31]:
# результат запроса
df = pd.read_sql_query(query_5_5, connection)
df

C:\Users\Maria\AppData\Local\Temp\ipykernel_4020\388711819.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


emp_name  count_industry
0        101 Интернет               4
1            21vek.by               4
2                2ГИС               4
3                  2К               4
4         4 пикселя +               4
...               ...             ...
1131            ЮРИОН               4
1132  ЮТИП Технологии               4
1133           ЯКласс               4
1134            ЯрНео               4
1135    Яшин&Партнёры               4

[1136 rows x 2 columns]

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [32]:
# текст запроса
query_5_6 = f'''SELECT COUNT(e.id)
                FROM public.employers e 
                    LEFT JOIN public.employers_industries ei ON e.id = ei.employer_id
                    LEFT JOIN public.industries i ON ei.industry_id = i.id
                WHERE i.name = 'Разработка программного обеспечения'
                '''

In [33]:
# результат запроса
df = pd.read_sql_query(query_5_6, connection)
df

C:\Users\Maria\AppData\Local\Temp\ipykernel_4020\821628714.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-миллионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [34]:
# код для получения списка городов-милионников
from bs4 import BeautifulSoup
import requests

response = requests.get('https://ru.wikipedia.org/wiki/Города-миллионеры_России') # запрашиваем содержимое веб-страницы
soup = BeautifulSoup(response.content)                                            # создаем объект BeautifulSoup, чтобы работать с содержимым
table = soup.select_one("table.standard:nth-child(11)")                           # выбираем нужную нам таблицу
body = table.select_one('tbody')                                                  # выбираем элемент таблицы, в котором находятся наши ячейки
cells = body.select('td')                                                         # создаем массив из всех ячеек таблицв поочереди

list = []                                                                         # создаем список из содержимого второго столбца таблицы
for i in range(len(cells)):
    if i % 5 == 1:
        list.append(cells[i])

cities = tuple([x.select_one('a').text for x in list])                            # получим кортеж из списка городов-миллионников
print(cities)

('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Челябинск', 'Красноярск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Омск', 'Краснодар', 'Воронеж', 'Пермь', 'Волгоград')


In [35]:
# текст запроса
query_5_7 = f'''(SELECT a.name, COUNT(e.name) cnt_vacancies
                FROM public.vacancies v 
                    LEFT JOIN public.employers e ON v.employer_id = e.id
                    LEFT JOIN public.areas a ON v.area_id = a.id

                WHERE e.name = 'Яндекс' AND a.name in {cities}
                GROUP BY a.name
                ORDER BY 2)

                UNION ALL
                (SELECT 'total', count(e.name) 
                FROM public.vacancies v 
                    LEFT JOIN public.employers e ON v.employer_id = e.id
                    LEFT JOIN public.areas a ON v.area_id = a.id

                WHERE e.name = 'Яндекс' AND a.name in {cities})
                ORDER BY 2
                '''

In [36]:
# результат запроса
df = pd.read_sql_query(query_5_7, connection)
df

C:\Users\Maria\AppData\Local\Temp\ipykernel_4020\1965596796.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


name  cnt_vacancies
0              Омск             21
1         Челябинск             22
2        Красноярск             23
3         Волгоград             24
4    Ростов-на-Дону             25
5             Пермь             25
6            Казань             25
7               Уфа             26
8            Самара             26
9         Краснодар             30
10          Воронеж             32
11      Новосибирск             35
12  Нижний Новгород             36
13     Екатеринбург             39
14  Санкт-Петербург             42
15           Москва             54
16            total            485

Результаты запросов:
1. По количеству предоставляемых вакансий на первом месте идет компания "Яндекс", следующая за ней компания - "Ростелеком", предоставляет даже не половину от количества вакансий "Яндекса". Это говорит нам о том, что компания "Яндекс" создает достаточно большое количество рабочих мест.
2. Среди регионов, в которых нет вакансий, наибольшее количество работодателей находится в регионе Россия.
3. При подсчете для каждого работодателя количество регионов, в которых он публикует свои вакансии, выясняется, что первое место опять занимает компания "Яндекс", она публикует свои вакансии в 181 регионе.
4. У 8419 работодателей не указана сфера деятельности. Скорее всего, это поле просто необязательное, поэтому работодатели ленятся его заполнять, или по названию вакансии можно прямо определить сферу деятельности.
5. У достаточно большого количества компаний (1136) указано 4 сферы деятельности. Можно предположить, что или компании требуется достаточно разносторонний специалист, или несколько специалистов в разных областях.
6. У 3553 работодателей в качестве сферы деятельности указана "Разработка программного обеспечения", это 15 % всех работодателей.
7. 25 % своих вакансий компания "Яндекс" размещает в городах-миллионниках.

По результам запросов компания "Яндекс" предоставляет огромное число вакансий, можно предположить, что такое число может быть связано с обширной деятельностью компании, в частности, компания занимается доставкой еды и предоставляет услуги такси. В принципе, крупные компании занимают всего 7 % от общего рынка вакансий. Среди работодателей в России (и других регионах) есть те, у которых нет свободных вакансий, это может говорить о том, что они уже нашли работника, но остались в базе, таких работодателей 410. Компания "Яндекс" является лидером не только по количеству вакансий, но и по размещению своих вакансий в 181 регионе, это опять свидетельствует об обширной деятельности компании. Большое количество указанных сфер деятельности может говорить о том, что компании требуется разностонний специалист. 15 % от общего количества вакансий имеют сферу деятельности "Разработка программного обеспечения" - это достаточно большой процент от общего числа вакансий, что говорит о востребованности данной сферы деятельности. Анализ последнего запроса показал, что 25 % своих вакансий компания "Яндекс" размещает в городах-миллионниках.

***

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [37]:
# текст запроса
query_6_1 = f'''SELECT COUNT(v.id)
                FROM public.vacancies v
                
                WHERE LOWER(v.name) LIKE '%data%' 
                    OR LOWER(v.name) LIKE '%данн%'
                '''

In [38]:
# результат запроса
df = pd.read_sql_query(query_6_1, connection)
df

C:\Users\Maria\AppData\Local\Temp\ipykernel_4020\366702258.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [39]:
# текст запроса
query_6_2 = f'''SELECT COUNT(v.id)
                FROM public.vacancies v
                
                WHERE (LOWER(v.name) LIKE '%data scientist%' 
                        OR LOWER(v.name) LIKE '%data science%' 
                        OR LOWER(v.name) LIKE '%исследователь данных%' 
                        OR (v.name like '%ML%' AND lower(v.name) NOT LIKE '%html%')
                        OR LOWER(v.name) LIKE '%machine learning%' 
                        OR LOWER(v.name) LIKE '%машинн%обучен%') 
                        AND (LOWER(v.name) LIKE '%junior%' 
                        OR v.experience LIKE '%Нет опыта%' 
                        OR v.employment LIKE '%Стажировка%')
                '''

In [40]:
# результат запроса
df = pd.read_sql_query(query_6_2, connection)
df

C:\Users\Maria\AppData\Local\Temp\ipykernel_4020\2267038862.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [41]:
# текст запроса
query_6_3 = f'''SELECT COUNT(id)
                FROM public.vacancies 
                
                WHERE (lower(name) LIKE '%data scientist%' 
                    OR lower(name) LIKE '%data science%'
                    OR lower(name) LIKE '%исследователь данных%' 
                    OR (name LIKE '%ML%' AND lower(name) NOT LIKE '%html%')
                    OR lower(name) LIKE '%machine learning%'
                    OR lower(name) LIKE '%машинн%обучен%')
                    AND (lower(key_skills)  LIKE '%sql%' 
                    OR lower(key_skills) LIKE '%postgres%')
                '''

In [42]:
# результат запроса
df = pd.read_sql_query(query_6_3, connection)
df

C:\Users\Maria\AppData\Local\Temp\ipykernel_4020\3794299702.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


count
0    201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [43]:
# текст запроса
query_6_4 = f'''SELECT COUNT(id)
                FROM public.vacancies
                
                WHERE (lower(name) LIKE '%data scientist%' 
                    OR lower(name) LIKE '%data science%'
                    OR lower(name) LIKE '%исследователь данных%' 
                    OR (name like '%ML%' AND lower(name) NOT LIKE '%html%') 
                    OR lower(name) LIKE '%machine learning%' 
                    OR lower(name) LIKE '%машинн%обучен%') 
                    AND (lower(key_skills)  LIKE '%python%')
                '''

In [44]:
# результат запроса
df = pd.read_sql_query(query_6_4, connection)
df

C:\Users\Maria\AppData\Local\Temp\ipykernel_4020\4264665179.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


count
0    351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [45]:
# текст запроса
query_6_5 = f'''SELECT
                    ROUND(AVG(LENGTH(key_skills) - LENGTH(REPLACE(key_skills, CHR(9), '')) + 1), 2) average
                FROM public.vacancies
            
                WHERE (lower(name) LIKE '%data scientist%' 
                    OR lower(name) LIKE '%data science%'
                    OR lower(name) LIKE '%исследователь данных%'
                    OR (name like '%ML%' AND lower(name) NOT LIKE '%html%')
                    OR lower(name) LIKE '%machine learning%'
                    OR lower(name) LIKE '%машинн%обучен%')
                    AND key_skills IS NOT NULL
                '''

In [46]:
# результат запроса
df = pd.read_sql_query(query_6_5, connection)
df

C:\Users\Maria\AppData\Local\Temp\ipykernel_4020\3654210517.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


average
0     6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [47]:
# текст запроса
query_6_6 = f'''SELECT
                experience,
                ROUND(AVG(COALESCE((salary_to + salary_from)/2, salary_to, salary_from))) average
            FROM public.vacancies
            
            WHERE (lower(name) LIKE '%data scientist%'
                OR lower(name) LIKE '%data science%' 
                OR lower(name) LIKE '%исследователь данных%' 
                OR (name LIKE '%ML%' AND lower(name) NOT LIKE '%html%')
                OR lower(name) LIKE '%machine learning%'
                OR lower(name) LIKE '%машинн%обучен%')
                AND (salary_from IS NOT NULL OR salary_to IS NOT NULL)
                AND (experience IS NOT NULL)
            
            GROUP BY experience
        '''

In [48]:
# результат запроса
df = pd.read_sql_query(query_6_6, connection)
df

C:\Users\Maria\AppData\Local\Temp\ipykernel_4020\519194770.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


experience   average
0           Нет опыта   74643.0
1  От 1 года до 3 лет  139675.0
2       От 3 до 6 лет  243115.0

Результаты запросов:
1. 1771 вакансия имеет отношение к данным.
2. Для начинающих дата-сайентистов имеется 51 вакансия.
3. SQL и postgres указаны в качестве ключевых навыков в 201 вакансии для DS.
4. Python указан в качестве ключевого навыка в 351 вакансии для DS.
5. 6.41 ключевых навыков в среднем указывают в вакансиях для DS.
6. По данному запросу можно сделать вывод, что чем больше опыт, тем выше зарплата, что очевидно.

Всего 3 % от общего количества вакансий, связанных с DS, предназначены для начинающих дата-сайентистов, это не так много, и говорит о желании работодателя найти опытного специалиста. А 11 % вакансий содержат в качестве ключевых навыков SQL и postgres, что свидетельствует о том, что эти навыки не так востребованы в этой сфере, но, на мой взгляд, могут послужить большим плюсом при устройстве на работу. В свою очередь, язык Python достаточно востребован для вакансий, связанных с DS, что говорит о актуальности данного языка программирования. Среднее число ключевых навыков для данных вакансий равно 6.41, т.е. специалист в области анализа данных - высококвалифицированные специалисты, обладающие большим количеством навыков. И последний пункт, это рост заработной платы пропорционально росту опыта работы, что достаточно очевидно. Можно отметить, что в среднем специалист с опытом работы "от 3 до 6 лет" получает зарплату в два раза больше, чем средняя верхняя вилка зарплаты, а начинающий специалист без опыта работы в среднем имеет доход больший, чем средняя нижняя вилка зарплаты. Такой уровень заработной платы сооответствует высокому уровню знаний, требуемых от рекрутера. 

***

# Общий вывод по проекту

In [49]:
# Выводим топ 10 по количеству вакансий среди сфер деятельности
query_7_1 = f'''SELECT i.name ind_name, COUNT(i.name)
                FROM public.employers e 
                    LEFT JOIN public.employers_industries ei ON e.id = ei.employer_id
                    LEFT JOIN public.industries i ON ei.industry_id = i.id
                GROUP BY i.name
                ORDER BY 2 DESC
                LIMIT 10
                '''

df = pd.read_sql_query(query_7_1, connection)
df

C:\Users\Maria\AppData\Local\Temp\ipykernel_4020\2970795094.py:11: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_1, connection)


ind_name  count
0                Разработка программного обеспечения   3553
1  Системная интеграция,  автоматизации технологи...   2993
2  Интернет-компания (поисковики, платежные систе...   1675
3  Маркетинговые, рекламные, BTL, дизайнерские, E...    798
4                              Консалтинговые услуги    662
5                                 Кадровые агентства    389
6  Электронно-вычислительная, оптическая, контрол...    374
7                       Розничная сеть (продуктовая)    368
8                                   Интернет-магазин    351
9                        Архитектура, проектирование    346

In [50]:
# Выводим топ по числу сочетаний навыков
query_7_2 = f'''SELECT key_skills, count(key_skills)
                FROM public.vacancies v
                GROUP BY key_skills
                ORDER BY 2 DESC
                LIMIT 10
                '''

df = pd.read_sql_query(query_7_2, connection)
df

C:\Users\Maria\AppData\Local\Temp\ipykernel_4020\2606496713.py:9: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_2, connection)


key_skills  count
0                                1С программирование     94
1  Пользователь ПК\tРабота в команде\tРемонт ноут...     83
2  PostgreSQL\tSQL\tLinux\t1С программирование\tM...     71
3          Ответственность\tКлиентоориентированность     48
4  MS PowerPoint\tПодготовка презентаций\tРазрабо...     27
5             1С программирование\t1С: Предприятие 8     26
6  Настройка ПК\tСборка ПК\tРемонт ПК\tНастройка ...     26
7                                    Английский язык     26
8  Грамотная речь\tРабота в команде\tГрамотность\...     25
9                                               Java     24

In [51]:
# Число вакансий с ключевым навыком Python
query_7_3 = f'''SELECT COUNT(id)
                FROM public.vacancies
                WHERE lower(key_skills)  LIKE '%python%'
                '''

df = pd.read_sql_query(query_7_3, connection)
df

C:\Users\Maria\AppData\Local\Temp\ipykernel_4020\339980770.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_3, connection)


count
0   3461

In [52]:
# Число вакансий с ключевым навыком Python и отсутствием опыта работы
query_7_4 = f'''SELECT COUNT(id)
                FROM public.vacancies
                WHERE lower(key_skills)  LIKE '%python%' AND experience = 'Нет опыта'
                '''

df = pd.read_sql_query(query_7_4, connection)
df

C:\Users\Maria\AppData\Local\Temp\ipykernel_4020\3519842405.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_4, connection)


count
0    236

In [53]:
# Среднее число ключевых навыков во всех вакансиях
query_7_5 = f'''SELECT
                    ROUND(AVG(LENGTH(key_skills) - LENGTH(REPLACE(key_skills, CHR(9), '')) + 1), 2) average
                FROM public.vacancies              
                '''

df = pd.read_sql_query(query_7_5, connection)
df

C:\Users\Maria\AppData\Local\Temp\ipykernel_4020\3670096567.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_5, connection)


average
0     6.37

In [54]:
# Число вакансий с некоторыми soft-skills навыками
query_7_6 = f'''(SELECT COUNT(id)
                FROM public.vacancies
                WHERE lower(key_skills) LIKE '%рамотная речь%') 
                
                UNION
                
                (SELECT COUNT(id)
                FROM public.vacancies
                WHERE lower(key_skills) LIKE '%абота в команде%')

                UNION
                (SELECT COUNT(id)
                FROM public.vacancies
                WHERE lower(key_skills) LIKE '%рганизационные навыки%')
                
                UNION
                (SELECT COUNT(id)
                FROM public.vacancies
                WHERE lower(key_skills) LIKE '%тветственность%')

                ORDER BY 1 DESC
                '''

df = pd.read_sql_query(query_7_6, connection)
df

C:\Users\Maria\AppData\Local\Temp\ipykernel_4020\1796072329.py:25: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_6, connection)


count
0   3924
1   2325
2    851
3    143

Вывод по дополнительным исследованиям:
1. Заметим, что первые места в топе 10 сфер деятельности по количеству вакансий, занимают сферы так или иначе связанные с IT, это говорит о высокой востребованности специалистов в этой области.
2. Данный запрос еще раз подчеркивает, что в топе по сочетанию нывыков первые места занимают именно навыки в сфере IT.
3. Всего 3461 вакансий с ключевым навыком Python - это 7 % от общего количества вакансий. Я сравнивала с остальными языками программирования, и Python - один из самых популярных и востребованных.
4. Среди вакансий с ключевым навыком Python есть 236 вакансий без требования к опыту работы, это 6.8 % от общего числа вакансий с ключевым навыком Python, к сожалению, даже в области программирования начинающему специалисту достаточно тяжело найти место работы.
5. Среднее число ключевых навыков во всех вакансиях равно 6.37, если сравнивать со средним числом ключевых навыков для специалиста DS (6.41), то оно практически не отличается, что говорит о том, что работодатели в целом хотят видеть в качестве соискателя человека с большим количеством навыков.
6. При рассмотрении вакансий с некоторыми soft-skills навыками было выявлено, что наиболее популярен навык "Работа в команде", так как в большинстве случаев, приходится работать в некой команде, и нужно уметь комуницировать с другими сотрудниками.

Общие выводы:

По данному исследованию базы данных можно сказать о том, что большинство вакансий распределено между столицами и крупными городами. Первое место с большим отрывом по количеству предоставляемых вакансий на рынке труда занимет компания "Яндекс", она является лидером не только по количеству вакансий, но и по размещению своих вакансий в 181 регионе, это свидетельствует об обширной деятельности компании (как подчеркивалось ранее). 25 % своих вакансий компания "Яндекс" размещает в городах-миллионниках.

Множество компаний в сфере своей деятельности указали "Разработка программного обеспечения", дополнительный анализ показал, что эта сфера деятельности занимает первое место по количеству вакансий. Если посмотреть на топ 10 сфер деятельности по количеству вакансий, то первые места занимают сферы так или иначе связанные с IT, это говорит о высокой востребованности специалистов в этой области. А также, если рассматривать число вакансий по сочетанию навыков, то первые места опять занимают вакансии в сфере IT, получается, что специалист в сфере IT - очень востребованный специалист, кандидат из этой сферы обязан иметь достаточно большое количество навыков, что накладывает высокие требования к подобному специалисту. Если говорить про языки программирования, то достаточно часто в вакансиях встречается Python.

Анализ вакансий DS показал, что всего 3 % от общего количества вакансий, предназначены для начинающих дата-сайентистов. Такие навыки как SQL и postgres могут послужить большим плюсом при устройстве на работу, а знание Python практически необходимо. Среднее число ключевых навыков для данных вакансий равно 6.41, т.е. специалист в области анализа данных - высококвалифицированные специалисты, обладающие большим количеством навыков. И последний пункт, это рост заработной платы пропорционально росту опыта работы, что достаточно очевидно. Можно отметить, что в среднем специалист с опытом работы "от 3 до 6 лет" получает зарплату в два раза больше, чем средняя верхняя вилка зарплаты, а начинающий специалист без опыта работы в среднем имеет доход больший, чем средняя нижняя вилка зарплаты. Такой уровень заработной платы сооответствует высокому уровню знаний, требуемых от рекрутера. Можно подчеркнуть, что на сегодняшний день начинающему специалисту достаточно непросто найти место работы, поэтому так важно иметь хорошее портфолио и достаточное число полезных навыков, чтобы конкурировать на рынке труда.

А при рассмотрении вакансий с некоторыми soft-skills навыками было выявлено, что наиболее популярен навык "Работа в команде", в принципе некоторые компании часто требуют тот или иной soft-skill.

In [55]:
# не забываем закрыть соединение после окончания работы
connection.close()